src/util/vector.h

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_UTIL_VECTOR_H
+#define BITCOIN_UTIL_VECTOR_H
+
+#include <vector>
+#include <type_traits>
+
+/** Emplace back multiple elements to a container. */
+template<typename V>
+inline void EmplaceBackMany(V& vec) {}
+template<typename V, typename Arg, typename... Args>
+inline void EmplaceBackMany(V& vec, Arg&& arg, Args&&... args)
+{
+    vec.emplace_back(std::forward<Arg>(arg));
+    EmplaceBackMany(vec, std::forward<Args>(args)...);
+}
+
+/** Construct a vector with the specified elements.
+ *
+ * This is preferable over the list initializing constructor of std::vector:
+ * - It automatically infers the element type from its arguments.
+ * - If any arguments are rvalue references, they will be moved into the vector
+ *   (list initialization always copies).
+ */
+template<typename... Args>
+inline std::vector<typename std::common_type<Args...>::type> Vector(Args&&... args)
+{
+    std::vector<typename std::common_type<Args...>::type> ret;
```

Comment:
Note to myself: https://en.cppreference.com/w/cpp/types/common_type says that the "types in the parameter pack T shall each be a complete type, [...]. Otherwise, the behavior is undefined". Though, I get a compilation failure when using incomplete types, which is nicer than undefined behavior.


```
/usr/include/c++/9/bits/stl_vector.h: In instantiation of ‘std::_Vector_base<_Tp, _Alloc>::~_Vector_base() [with _Tp = A; _Alloc = std::allocator<A>]’:
/usr/include/c++/9/bits/stl_vector.h:484:7:   required from ‘std::vector<typename std::common_type<_Tp>::type> Vector(Args&& ...) [with Args = {A&}; typename std::common_type<_Tp>::type = A]’
1.cpp:81:12:   required from here
/usr/include/c++/9/bits/stl_vector.h:333:35: error: invalid use of incomplete type ‘class A’
  333 |         _M_impl._M_end_of_storage - _M_impl._M_start);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
1.cpp:77:7: note: forward declaration of ‘class A’
   77 | class A;
      |       ^

Reply:
Given that a vector is actually instantiated of the resulting type, I don't think there is much undefined "compile time" behavior that can lead to actually compiling code. There is also essentially just one type that vector can be; if it's wrong, the elements won't be able to be added to it.